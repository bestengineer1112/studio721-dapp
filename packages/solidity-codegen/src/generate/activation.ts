import {
  AST,
  blockComment,
  functionDeclaration,
  identifierExpression,
  literalExpression,
  variableDeclaration,
} from 'solidity-language';

export function generateActivation({
  initialValue,
  toggleAccessToken,
  mutableAccessToken,
}: {
  initialValue: boolean;
  toggleAccessToken: boolean;
  mutableAccessToken: boolean;
}): (AST.Declaration | AST.BlockComment)[] {
  return [
    blockComment({ value: 'ACTIVATION', commentType: '/*' }),
    variableDeclaration({
      name: 'saleIsActive',
      typeAnnotation: 'bool',
      modifiers: ['public'],
      initializer: literalExpression(String(initialValue)),
    }),
    functionDeclaration({
      name: 'setSaleIsActive',
      arguments: ['bool saleIsActive_'],
      modifiers: ['external', 'onlyOwner'],
      body: [
        {
          type: 'expressionStatement',
          expression: {
            type: 'assignmentExpression',
            lhs: identifierExpression('saleIsActive'),
            rhs: identifierExpression('saleIsActive_'),
          },
        },
      ],
    }),
    ...(toggleAccessToken
      ? [
          variableDeclaration({
            name: 'accessTokenIsActive',
            typeAnnotation: 'bool',
            modifiers: ['public'],
            initializer: literalExpression(String(true)),
          }),
          functionDeclaration({
            name: 'setAccessTokenIsActive',
            arguments: ['bool accessTokenIsActive_'],
            modifiers: ['external', 'onlyOwner'],
            body: [
              {
                type: 'expressionStatement',
                expression: {
                  type: 'assignmentExpression',
                  lhs: identifierExpression('accessTokenIsActive'),
                  rhs: identifierExpression('accessTokenIsActive_'),
                },
              },
            ],
          }),
        ]
      : []),
    ...(mutableAccessToken
      ? [
          functionDeclaration({
            name: 'setAccessTokenAddress',
            arguments: ['address accessTokenAddress_'],
            modifiers: ['external', 'onlyOwner'],
            body: [
              {
                type: 'expressionStatement',
                expression: {
                  type: 'assignmentExpression',
                  lhs: identifierExpression('accessTokenAddress'),
                  rhs: identifierExpression('accessTokenAddress_'),
                },
              },
            ],
          }),
        ]
      : []),
  ];
}
