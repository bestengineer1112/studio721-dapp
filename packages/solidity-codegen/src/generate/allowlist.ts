import { ContractConfigState } from 'state';
import {
  AST,
  binaryExpression,
  blockComment,
  expressionStatement,
  functionCallExpression,
  functionDeclaration,
  identifierExpression,
  ifStatement,
  parseVariableDeclaration,
  returnStatement,
  variableDeclaration,
} from 'solidity-language';

export function generateAllowlist(
  config: Pick<
    ContractConfigState,
    'limitPerWallet' | 'amountAllowedForOwner' | 'allowlistDestinations'
  >,
): (AST.Declaration | AST.BlockComment)[] {
  const hasAllowlist =
    config.amountAllowedForOwner > 0 || config.allowlistDestinations.length > 0;

  if (!hasAllowlist && config.limitPerWallet === undefined) {
    return [];
  }

  return [
    blockComment({ value: 'MINTING LIMITS', commentType: '/*' }),
    variableDeclaration({
      name: 'mintCountMap',
      modifiers: ['private'],
      typeAnnotation: 'mapping(address => uint256)',
    }),
    variableDeclaration({
      name: 'allowedMintCountMap',
      modifiers: ['private'],
      typeAnnotation: 'mapping(address => uint256)',
    }),
    ...(config.limitPerWallet !== undefined
      ? [
          parseVariableDeclaration(
            `uint256 public constant MINT_LIMIT_PER_WALLET = ${config.limitPerWallet}`,
          ),
        ]
      : []),
    ...(hasAllowlist && config.limitPerWallet !== undefined
      ? [
          functionDeclaration({
            name: 'max',
            arguments: ['uint256 a', 'uint256 b'],
            modifiers: ['private pure'],
            returns: {
              modifiers: [],
              typeAnnotation: 'uint256',
            },
            body: [returnStatement(identifierExpression('a >= b ? a : b'))],
          }),
        ]
      : []),
    functionDeclaration({
      name: 'allowedMintCount',
      arguments: ['address minter'],
      modifiers: ['public', 'view'],
      returns: {
        modifiers: [],
        typeAnnotation: 'uint256',
      },
      body: [
        ...(hasAllowlist && config.limitPerWallet !== undefined
          ? [
              ifStatement({
                condition: identifierExpression('saleIsActive'),
                body: [
                  returnStatement(
                    binaryExpression({
                      lhs: functionCallExpression({
                        callee: identifierExpression('max'),
                        arguments: [
                          identifierExpression('allowedMintCountMap[minter]'),
                          identifierExpression('MINT_LIMIT_PER_WALLET'),
                        ],
                      }),
                      operator: '-',
                      rhs: identifierExpression('mintCountMap[minter]'),
                    }),
                  ),
                ],
              }),
            ]
          : []),
        ...(hasAllowlist
          ? [
              returnStatement(
                binaryExpression({
                  lhs: identifierExpression('allowedMintCountMap[minter]'),
                  operator: '-',
                  rhs: identifierExpression('mintCountMap[minter]'),
                }),
              ),
            ]
          : [
              returnStatement(
                binaryExpression({
                  lhs: identifierExpression('MINT_LIMIT_PER_WALLET'),
                  operator: '-',
                  rhs: identifierExpression('mintCountMap[minter]'),
                }),
              ),
            ]),
      ],
    }),
    functionDeclaration({
      name: 'updateMintCount',
      arguments: ['address minter', 'uint256 count'],
      modifiers: ['private'],
      body: [
        expressionStatement({
          expression: binaryExpression({
            lhs: identifierExpression('mintCountMap[minter]'),
            operator: '+=',
            rhs: identifierExpression('count'),
          }),
        }),
      ],
    }),
  ];
}
