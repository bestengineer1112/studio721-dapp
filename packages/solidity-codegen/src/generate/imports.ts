import { AST } from 'solidity-language';

export function generateImports({
  address,
  strings,
  enumerable,
  royalties,
  delegation,
  hasAccessToken,
}: {
  address: boolean;
  strings: boolean;
  enumerable: boolean;
  royalties: boolean;
  delegation: boolean;
  hasAccessToken: boolean;
}): AST.Import[] {
  const paths: (string | AST.Import)[] = [
    ...(delegation
      ? [
          {
            path: 'gwei-slim-nft-contracts/contracts/base/ERC721Base.sol',
            names: ['ConfigSettings'],
          },
          {
            path: 'gwei-slim-nft-contracts/contracts/base/ERC721Delegated.sol',
            names: ['ERC721Delegated'],
          },
          ...(hasAccessToken
            ? ['@openzeppelin/contracts/token/ERC721/IERC721.sol']
            : []),
        ]
      : [
          '@openzeppelin/contracts/token/ERC721/ERC721.sol',
          '@openzeppelin/contracts/access/Ownable.sol',
        ]),
    '@openzeppelin/contracts/security/ReentrancyGuard.sol',
    ...(enumerable
      ? ['@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol']
      : []),
    ...(royalties ? ['@openzeppelin/contracts/interfaces/IERC2981.sol'] : []),
    '@openzeppelin/contracts/utils/Counters.sol',
    ...(address ? ['@openzeppelin/contracts/utils/Address.sol'] : []),
    ...(strings ? ['@openzeppelin/contracts/utils/Strings.sol'] : []),
  ];

  return paths.map((pathOrImport) => {
    if (typeof pathOrImport === 'string') {
      return { path: pathOrImport, names: [] };
    } else {
      return pathOrImport;
    }
  });
}
