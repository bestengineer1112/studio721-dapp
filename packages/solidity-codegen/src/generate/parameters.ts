import { ContractConfigState } from 'state';
import {
  AST,
  blockComment,
  functionDeclaration,
  structDeclaration,
  structMemberDeclaration,
  variableDeclaration,
} from 'solidity-language';

export function generateTokenParameters(
  config: Pick<ContractConfigState, 'tokenParameters'>,
): (AST.Declaration | AST.BlockComment)[] {
  return [
    blockComment({ value: 'TOKEN PARAMETERS', commentType: '/*' }),
    structDeclaration({
      name: 'TokenParameters',
      body: config.tokenParameters.map((parameter) =>
        structMemberDeclaration({
          name: parameter.name,
          typeAnnotation: parameter.type,
        }),
      ),
    }),
    variableDeclaration({
      name: 'tokenParametersMap',
      modifiers: ['private'],
      typeAnnotation: 'mapping(uint256 => TokenParameters)',
    }),
    functionDeclaration({
      name: 'tokenParameters',
      arguments: ['uint256 tokenId'],
      modifiers: ['external', 'view'],
      returns: {
        modifiers: ['memory'],
        typeAnnotation: 'TokenParameters',
      },
      body: [
        {
          type: 'returnStatement',
          expression: {
            type: 'identifier',
            value: 'tokenParametersMap[tokenId]',
          },
        },
      ],
    }),
  ];
}
