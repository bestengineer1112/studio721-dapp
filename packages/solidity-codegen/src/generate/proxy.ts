import {
  AST,
  blockComment,
  contractDeclaration,
  declarationStatement,
  functionCallExpression,
  functionDeclaration,
  identifierExpression,
  ifStatement,
  literalExpression,
  parseVariableDeclaration,
  returnStatement,
  variableDeclaration,
} from 'solidity-language';

export function generateProxyContracts(): AST.ContractDeclaration[] {
  return [
    contractDeclaration({
      name: 'OwnableDelegateProxy',
    }),
    contractDeclaration({
      name: 'ProxyRegistry',
      body: [
        variableDeclaration({
          name: 'proxies',
          modifiers: ['public'],
          typeAnnotation: 'mapping(address => OwnableDelegateProxy)',
        }),
      ],
    }),
  ];
}

export function generateProxyApprovalFunction({
  usesDelegatedContract,
}: {
  usesDelegatedContract: boolean;
}) {
  return [
    blockComment({ value: 'PROXY REGISTRY', commentType: '/*' }),
    parseVariableDeclaration('address private immutable proxyRegistryAddress'),
    functionDeclaration({
      name: 'isApprovedForAll',
      arguments: ['address owner', 'address operator'],
      modifiers: [
        ...(usesDelegatedContract ? [] : ['override']),
        'public',
        'view',
      ],
      returns: {
        typeAnnotation: 'bool',
        modifiers: [],
      },
      body: [
        declarationStatement({
          declaration: variableDeclaration({
            name: 'proxyRegistry',
            typeAnnotation: 'ProxyRegistry',
            initializer: functionCallExpression({
              callee: identifierExpression('ProxyRegistry'),
              arguments: [identifierExpression('proxyRegistryAddress')],
            }),
          }),
        }),
        ifStatement({
          condition: identifierExpression(
            'address(proxyRegistry.proxies(owner)) == operator',
          ),
          body: [returnStatement(literalExpression('true'))],
        }),
        returnStatement(
          functionCallExpression({
            callee: usesDelegatedContract
              ? identifierExpression('_isApprovedForAll')
              : identifierExpression('super.isApprovedForAll'),
            arguments: [
              identifierExpression('owner'),
              identifierExpression('operator'),
            ],
          }),
        ),
      ],
    }),
  ];
}
