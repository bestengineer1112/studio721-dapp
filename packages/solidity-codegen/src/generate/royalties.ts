import {
  AST,
  binaryExpression,
  blockComment,
  functionDeclaration,
  identifierExpression,
  returnStatement,
} from 'solidity-language';

export function generateRoyalties(config: {
  royaltyBps: string;
}): (AST.Declaration | AST.BlockComment)[] {
  return [
    blockComment({ value: 'ROYALTIES', commentType: '/*' }),
    functionDeclaration({
      name: 'royaltyInfo',
      arguments: ['uint256', 'uint256 salePrice'],
      modifiers: ['external', 'view', 'override'],
      returns: {
        modifiers: [],
        typeAnnotation: 'address receiver, uint256 royaltyAmount',
      },
      body: [
        returnStatement(
          identifierExpression(
            `(address(this), (salePrice * ${
              Number(config.royaltyBps) * 100
            }) / 10000)`,
          ),
        ),
      ],
    }),
    functionDeclaration({
      name: 'supportsInterface',
      arguments: ['bytes4 interfaceId'],
      modifiers: ['public', 'view', 'virtual', 'override(ERC721, IERC165)'],
      returns: {
        modifiers: [],
        typeAnnotation: 'bool',
      },
      body: [
        returnStatement(
          binaryExpression({
            operator: '||',
            lhs: identifierExpression(
              'interfaceId == type(IERC2981).interfaceId',
            ),
            rhs: identifierExpression('super.supportsInterface(interfaceId)'),
          }),
        ),
      ],
    }),
  ];
}
