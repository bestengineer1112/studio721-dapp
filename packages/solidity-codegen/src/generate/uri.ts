import { ContractConfigState } from 'state';
import {
  assignmentExpression,
  AST,
  blockComment,
  declarationStatement,
  expressionStatement,
  functionCallExpression,
  functionDeclaration,
  identifierExpression,
  ifStatement,
  literalExpression,
  parseVariableDeclaration,
  returnStatement,
  variableDeclaration,
} from 'solidity-language';
import { parseURITemplate } from '../templateParser';

function getURIConfig(baseURI: string) {
  const tokenizedURI = parseURITemplate(baseURI);
  const tokens = tokenizedURI.type === 'success' ? tokenizedURI.value : [];

  const hasBaseURI = tokens.length > 0 && tokens[0].type === 'content';
  const isDefaultConfiguration =
    (tokens.length === 2 &&
      tokens[0].type === 'content' &&
      tokens[1].type === 'identifier' &&
      tokens[1].values[0] === 'tokenId') ||
    (tokens.length === 3 &&
      tokens[0].type === 'content' &&
      tokens[1].type === 'identifier' &&
      tokens[1].values[0] === 'tokenId' &&
      tokens[2].type === 'identifier' &&
      tokens[2].values[0] === 'parameters');
  const startsWithDefaultConfiguration =
    tokens.length >= 2 &&
    tokens[0].type === 'content' &&
    tokens[1].type === 'identifier' &&
    tokens[1].values[0] === 'tokenId';
  const usesParameters = tokens.some(
    (token) =>
      token.type === 'identifier' && token.values.includes('parameters'),
  );

  return {
    tokens,
    isDefaultConfiguration,
    startsWithDefaultConfiguration,
    hasBaseURI,
    usesParameters,
  };
}

function generateContractURI(uri: string) {
  return [
    parseVariableDeclaration(
      `string private customContractURI = ${JSON.stringify(encodeURI(uri))}`,
    ),
    functionDeclaration({
      name: 'setContractURI',
      arguments: ['string memory customContractURI_'],
      modifiers: ['external', 'onlyOwner'],
      body: [
        expressionStatement({
          expression: assignmentExpression({
            lhs: identifierExpression('customContractURI'),
            rhs: identifierExpression('customContractURI_'),
          }),
        }),
      ],
    }),
    functionDeclaration({
      name: 'contractURI',
      modifiers: ['public', 'view'],
      returns: { typeAnnotation: 'string', modifiers: ['memory'] },
      body: [returnStatement(identifierExpression('customContractURI'))],
    }),
  ];
}

export function generateURIHandling(
  config: Pick<
    ContractConfigState,
    | 'tokenParameters'
    | 'tokenURI'
    | 'usesUriStorage'
    | 'usesDelegatedContract'
    | 'contractURI'
  >,
): (AST.Declaration | AST.BlockComment)[] {
  const {
    tokens,
    isDefaultConfiguration,
    startsWithDefaultConfiguration,
    usesParameters,
  } = getURIConfig(config.tokenURI);

  const shouldReturnSimpleBaseURI =
    tokens.length === 1 ||
    (tokens.length === 2 &&
      usesParameters &&
      config.tokenParameters.length === 0);

  return [
    blockComment({ value: 'URI HANDLING', commentType: '/*' }),
    ...(config.contractURI ? generateContractURI(config.contractURI) : []),
    ...(!config.usesDelegatedContract
      ? [parseVariableDeclaration('string private customBaseURI')]
      : []),
    ...(config.usesUriStorage
      ? [
          variableDeclaration({
            name: 'tokenURIMap',
            modifiers: ['private'],
            typeAnnotation: 'mapping(uint256 => string)',
          }),
          functionDeclaration({
            name: 'setTokenURI',
            arguments: ['uint256 tokenId', 'string memory tokenURI_'],
            modifiers: ['external', 'onlyOwner'],
            body: [
              {
                type: 'expressionStatement',
                expression: {
                  type: 'assignmentExpression',
                  lhs: {
                    type: 'identifier',
                    value: 'tokenURIMap[tokenId]',
                  },
                  rhs: {
                    type: 'identifier',
                    value: 'tokenURI_',
                  },
                },
              },
            ],
          }),
        ]
      : []),
    functionDeclaration({
      name: 'setBaseURI',
      arguments: ['string memory customBaseURI_'],
      modifiers: ['external', 'onlyOwner'],
      body: [
        config.usesDelegatedContract
          ? expressionStatement({
              expression: functionCallExpression({
                callee: identifierExpression('_setBaseURI'),
                arguments: [
                  identifierExpression('customBaseURI_'),
                  literalExpression('""'),
                ],
              }),
            })
          : expressionStatement({
              expression: assignmentExpression({
                lhs: identifierExpression('customBaseURI'),
                rhs: identifierExpression('customBaseURI_'),
              }),
            }),
      ],
    }),
    ...(!config.usesDelegatedContract
      ? [
          functionDeclaration({
            name: '_baseURI',
            arguments: [],
            modifiers: [
              'internal',
              'view',
              'virtual',
              ...(config.usesDelegatedContract ? [] : ['override']),
            ],
            returns: { typeAnnotation: 'string', modifiers: ['memory'] },
            body: [
              {
                type: 'returnStatement',
                expression: {
                  type: 'identifier',
                  value: 'customBaseURI',
                },
              },
            ],
          }),
        ]
      : []),
    ...(!isDefaultConfiguration ||
    (usesParameters && config.tokenParameters.length > 0) ||
    config.usesUriStorage
      ? [
          functionDeclaration({
            name: 'tokenURI',
            arguments:
              // We need to include the parameter, even when we don't use it,
              // so that the function has the right signature
              shouldReturnSimpleBaseURI && !config.usesUriStorage
                ? ['uint256']
                : ['uint256 tokenId'],
            modifiers: [
              'public',
              'view',
              ...(config.usesDelegatedContract ? [] : ['override']),
            ],
            returns: { typeAnnotation: 'string', modifiers: ['memory'] },
            body: [
              ...(config.usesUriStorage
                ? [
                    declarationStatement({
                      declaration: parseVariableDeclaration(
                        'string memory tokenURI_ = tokenURIMap[tokenId]',
                      ),
                    }),
                    ifStatement({
                      condition: identifierExpression(
                        'bytes(tokenURI_).length > 0',
                      ),
                      body: [
                        returnStatement(identifierExpression('tokenURI_')),
                      ],
                    }),
                  ]
                : []),
              ...(usesParameters && config.tokenParameters.length > 0
                ? [
                    declarationStatement({
                      declaration: variableDeclaration({
                        name: 'parameters',
                        modifiers: ['memory'],
                        typeAnnotation: 'TokenParameters',
                        initializer: identifierExpression(
                          'tokenParametersMap[tokenId]',
                        ),
                      }),
                    }),
                  ]
                : []),
              {
                type: 'returnStatement',
                expression: shouldReturnSimpleBaseURI
                  ? functionCallExpression({
                      callee: identifierExpression('_baseURI'),
                    })
                  : functionCallExpression({
                      callee: identifierExpression('string'),
                      arguments: [
                        functionCallExpression({
                          callee: identifierExpression('abi.encodePacked'),
                          arguments: [
                            ...tokens.flatMap(
                              (token, index): AST.Expression[] => {
                                if (startsWithDefaultConfiguration) {
                                  if (index === 0) {
                                    return [
                                      identifierExpression(
                                        config.usesDelegatedContract
                                          ? '_tokenURI(tokenId)'
                                          : 'super.tokenURI(tokenId)',
                                      ),
                                    ];
                                  } else if (index === 1) {
                                    return [];
                                  }
                                }

                                switch (token.type) {
                                  case 'content':
                                    return [
                                      identifierExpression(
                                        JSON.stringify(token.values.join('')),
                                      ),
                                    ];
                                  case 'identifier': {
                                    const values = token.values[0];

                                    switch (values) {
                                      case 'tokenId':
                                        return [
                                          identifierExpression('tokenId'),
                                        ];
                                      case 'parameters':
                                        if (config.tokenParameters.length === 0)
                                          return [];

                                        return [
                                          identifierExpression('"?"'),
                                          ...intersperse(
                                            config.tokenParameters.flatMap(
                                              (parameter) => [
                                                identifierExpression(
                                                  `"${parameter.name}="`,
                                                ),
                                                identifierExpression(
                                                  parameter.type === 'string' ||
                                                    parameter.type === 'address'
                                                    ? `parameters.${parameter.name}`
                                                    : `parameters.${parameter.name}.toString()`,
                                                ),
                                              ],
                                            ),
                                            identifierExpression('"&"'),
                                            2,
                                          ),
                                        ];
                                      default:
                                        return [
                                          identifierExpression(
                                            JSON.stringify(
                                              token.values.join(''),
                                            ),
                                          ),
                                        ];
                                    }
                                  }
                                  default:
                                    throw new Error('Invalid');
                                }
                              },
                            ),
                          ],
                        }),
                      ],
                    }),
              },
            ],
          }),
        ]
      : []),
  ];
}

export default function intersperse<T>(
  elements: T[],
  separator: T,
  stride = 1,
) {
  return elements.flatMap((element, i) =>
    i === 0 || i % stride !== 0 ? [element] : [separator, element],
  );
}
