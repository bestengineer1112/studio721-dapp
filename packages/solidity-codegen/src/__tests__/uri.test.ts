import { generateURIHandling } from '../generate/uri';
import { contractDeclaration, formatContract, print } from 'solidity-language';

describe('uri handling', () => {
  it('base uri only', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/',
        tokenParameters: [],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base uri only + unused parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{parameters}',
        tokenParameters: [],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base uri and token id', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}',
        tokenParameters: [],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base uri + token id + suffix', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}.json',
        tokenParameters: [],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('token parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}{parameters}',
        tokenParameters: [{ name: 'a', type: 'uint256' }],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multiple token parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}{parameters}',
        tokenParameters: [
          { name: 'a', type: 'uint256' },
          { name: 'b', type: 'uint256' },
        ],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multiple token parameters + file suffix', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}.json{parameters}',
        tokenParameters: [
          { name: 'a', type: 'uint256' },
          { name: 'b', type: 'uint256' },
        ],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('unused parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}',
        tokenParameters: [{ name: 'a', type: 'uint256' }],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('usesUriStorage', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: true,
        tokenURI: 'https://example.com/',
        tokenParameters: [],
        usesDelegatedContract: false,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('delegation', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}',
        tokenParameters: [],
        usesDelegatedContract: true,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('delegation + multiple token parameters + file suffix', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}.json{parameters}',
        tokenParameters: [
          { name: 'a', type: 'uint256' },
          { name: 'b', type: 'uint256' },
        ],
        usesDelegatedContract: true,
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('contract uri', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateURIHandling({
        usesUriStorage: false,
        tokenURI: 'https://example.com/{tokenId}',
        tokenParameters: [],
        usesDelegatedContract: false,
        contractURI: 'https://example.com/contract.json',
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });
});
