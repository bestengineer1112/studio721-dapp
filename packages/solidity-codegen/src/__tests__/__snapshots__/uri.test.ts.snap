// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uri handling base uri + token id + suffix 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }

  function tokenURI(uint256 tokenId) public view override
    returns (string memory)
  {
    return string(abi.encodePacked(super.tokenURI(tokenId), \\".json\\"));
  }
}"
`;

exports[`uri handling base uri and token id 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }
}"
`;

exports[`uri handling base uri only + unused parameters 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }

  function tokenURI(uint256) public view override returns (string memory) {
    return _baseURI();
  }
}"
`;

exports[`uri handling base uri only 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }

  function tokenURI(uint256) public view override returns (string memory) {
    return _baseURI();
  }
}"
`;

exports[`uri handling contract uri 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customContractURI = \\"https://example.com/contract.json\\";

  function setContractURI(string memory customContractURI_) external onlyOwner {
    customContractURI = customContractURI_;
  }

  function contractURI() public view returns (string memory) {
    return customContractURI;
  }

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }
}"
`;

exports[`uri handling delegation + multiple token parameters + file suffix 1`] = `
"contract Test {
  /** URI HANDLING **/

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    _setBaseURI(customBaseURI_, \\"\\");
  }

  function tokenURI(uint256 tokenId) public view returns (string memory) {
    TokenParameters memory parameters = tokenParametersMap[tokenId];

    return (
      string(
        abi.encodePacked(
          _tokenURI(tokenId),
          \\".json\\",
          \\"?\\",
          \\"a=\\",
          parameters.a.toString(),
          \\"&\\",
          \\"b=\\",
          parameters.b.toString()
        )
      )
    );
  }
}"
`;

exports[`uri handling delegation 1`] = `
"contract Test {
  /** URI HANDLING **/

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    _setBaseURI(customBaseURI_, \\"\\");
  }
}"
`;

exports[`uri handling multiple token parameters + file suffix 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }

  function tokenURI(uint256 tokenId) public view override
    returns (string memory)
  {
    TokenParameters memory parameters = tokenParametersMap[tokenId];

    return (
      string(
        abi.encodePacked(
          super.tokenURI(tokenId),
          \\".json\\",
          \\"?\\",
          \\"a=\\",
          parameters.a.toString(),
          \\"&\\",
          \\"b=\\",
          parameters.b.toString()
        )
      )
    );
  }
}"
`;

exports[`uri handling multiple token parameters 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }

  function tokenURI(uint256 tokenId) public view override
    returns (string memory)
  {
    TokenParameters memory parameters = tokenParametersMap[tokenId];

    return (
      string(
        abi.encodePacked(
          super.tokenURI(tokenId),
          \\"?\\",
          \\"a=\\",
          parameters.a.toString(),
          \\"&\\",
          \\"b=\\",
          parameters.b.toString()
        )
      )
    );
  }
}"
`;

exports[`uri handling token parameters 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }

  function tokenURI(uint256 tokenId) public view override
    returns (string memory)
  {
    TokenParameters memory parameters = tokenParametersMap[tokenId];

    return (
      string(
        abi.encodePacked(
          super.tokenURI(tokenId),
          \\"?\\",
          \\"a=\\",
          parameters.a.toString()
        )
      )
    );
  }
}"
`;

exports[`uri handling unused parameters 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }
}"
`;

exports[`uri handling usesUriStorage 1`] = `
"contract Test {
  /** URI HANDLING **/

  string private customBaseURI;

  mapping(uint256 => string) private tokenURIMap;

  function setTokenURI(uint256 tokenId, string memory tokenURI_)
    external
    onlyOwner
  {
    tokenURIMap[tokenId] = tokenURI_;
  }

  function setBaseURI(string memory customBaseURI_) external onlyOwner {
    customBaseURI = customBaseURI_;
  }

  function _baseURI() internal view virtual override returns (string memory) {
    return customBaseURI;
  }

  function tokenURI(uint256 tokenId) public view override
    returns (string memory)
  {
    string memory tokenURI_ = tokenURIMap[tokenId];

    if (bytes(tokenURI_).length > 0) {
      return tokenURI_;
    }

    return _baseURI();
  }
}"
`;
